### ISSUE TITLE
yum not idempotent when rpm already installed

### ISSUE THREAD

**Issue by chris-kiick-sp** [ansible/ansible#71758]
##### SUMMARY

Using yum task on a URL pointing to rpm will fail if newer version already installed.  Yum reports "nothing to do".
Update: the  already installed package must be newer version than the rpm.
##### ISSUE TYPE
- Bug Report

##### COMPONENT NAME

yum

##### ANSIBLE VERSION

```paste below
ansible 2.9.11
  config file = None
  configured module search path = [u'/home/chris.kiick/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /usr/lib/python2.7/site-packages/ansible
  executable location = /usr/bin/ansible
  python version = 2.7.5 (default, Jun 11 2019, 14:33:56) [GCC 4.8.5 20150623 (Red Hat 4.8.5-39)]
```

##### CONFIGURATION

```paste below

```

##### OS / ENVIRONMENT

Red Hat Enterprise Linux Server release 7.8 (Maipo)
Running on AWS EC2 instance in us-gov-west-1
NOTE: FIPS mode is enabled.

##### STEPS TO REPRODUCE

Install amazon ssm agent, then have it update itself (can be done in AWS console)
Run ansible-playbook on the below yaml.

```yaml
- hosts: localhost
  tasks:
    - yum:
        name: https://s3.amazonaws.com/ec2-downloads-windows/SSMAgent/latest/linux_amd64/amazon-ssm-agent.rpm
        state: latest
```

##### EXPECTED RESULTS

Should report "OK" with changed set to false.

##### ACTUAL RESULTS

```paste below
TASK [yum] *********************************************************************
task path: /home/chris.kiick/tp.yml:3
<127.0.0.1> ESTABLISH LOCAL CONNECTION FOR USER: root
<127.0.0.1> EXEC /bin/sh -c 'echo ~root && sleep 0'
<127.0.0.1> EXEC /bin/sh -c '( umask 77 && mkdir -p "` echo /root/.ansible/tmp `"&& mkdir /root/.ansible/tmp/ansible-tmp-1600178171.19-5797-112509477499552 && echo ansible-tmp-1600178171.19-5797-112509477499552="` echo /root/.ansible/tmp/ansible-tmp-1600178171.19-5797-112509477499552 `" ) && sleep 0'
Using module file /usr/lib/python2.7/site-packages/ansible/modules/packaging/os/yum.py
<127.0.0.1> PUT /root/.ansible/tmp/ansible-local-5741OK9f3Q/tmp5CTAyK TO /root/.ansible/tmp/ansible-tmp-1600178171.19-5797-112509477499552/AnsiballZ_yum.py
<127.0.0.1> EXEC /bin/sh -c 'chmod u+x /root/.ansible/tmp/ansible-tmp-1600178171.19-5797-112509477499552/ /root/.ansible/tmp/ansible-tmp-1600178171.19-5797-112509477499552/AnsiballZ_yum.py && sleep 0'
<127.0.0.1> EXEC /bin/sh -c '/usr/bin/python2 /root/.ansible/tmp/ansible-tmp-1600178171.19-5797-112509477499552/AnsiballZ_yum.py && sleep 0'
<127.0.0.1> EXEC /bin/sh -c 'rm -f -r /root/.ansible/tmp/ansible-tmp-1600178171.19-5797-112509477499552/ > /dev/null 2>&1 && sleep 0'
fatal: [localhost]: FAILED! => {
    "changed": true,
    "changes": {
        "installed": [
            "https://s3.amazonaws.com/ec2-downloads-windows/SSMAgent/latest/linux_amd64/amazon-ssm-agent.rpm"
        ],
        "updated": []
    },
    "invocation": {
        "module_args": {
            "allow_downgrade": false,
            "autoremove": false,
            "bugfix": false,
            "conf_file": null,
            "disable_excludes": null,
            "disable_gpg_check": false,
            "disable_plugin": [],
            "disablerepo": [],
            "download_dir": null,
            "download_only": false,
            "enable_plugin": [],
            "enablerepo": [],
            "exclude": [],
            "install_repoquery": true,
            "install_weak_deps": true,
            "installroot": "/",
            "list": null,
            "lock_timeout": 30,
            "name": [
                "https://s3.amazonaws.com/ec2-downloads-windows/SSMAgent/latest/linux_amd64/amazon-ssm-agent.rpm"
            ],
            "releasever": null,
            "security": false,
            "skip_broken": false,
            "state": "latest",
            "update_cache": false,
            "update_only": false,
            "use_backend": "auto",
            "validate_certs": true
        }
    },
    "msg": "Error: Nothing to do\n",
    "obsoletes": {
        "iwl7265-firmware": {
            "dist": "noarch",
            "repo": "@anaconda/7.7",
            "version": "22.0.7.0-72.el7"
        }
    },
    "rc": 1,
    "results": [
        "Loaded plugins: amazon-id, search-disabled-repos\nExamining /var/tmp/yum-root-IuL3gM/amazon-ssm-agent.rpm: amazon-ssm-agent-2.3.1644.0-1.x86_64\n/var/tmp/yum-root-IuL3gM/amazon-ssm-agent.rpm: does not update installed package.\n"
    ]
}

PLAY RECAP *********************************************************************
localhost                  : ok=1    changed=0    unreachable=0    failed=1    skipped=0    rescued=0    ignored=0
```

**Comment by chris-kiick-sp** [comment]
Not sure if this is the same bug: https://github.com/ansible/ansible/issues/49177

**Comment by chris-kiick-sp** [comment]
More info:
The package already installed on the host was a newer version than the one in the rpm.  I would expect in this case that the "latest" flag would Do The Right Thing and update the package if the rpm was a newer version.  Using the "present" flag would not handle the case where the rpm was newer (at least according to the documentation).

# sudo  yum info amazon-ssm-agent
Loaded plugins: amazon-id, search-disabled-repos
Installed Packages
Name        : amazon-ssm-agent
Arch        : x86_64
Version     : 3.0.161.0
Release     : 1
Size        : 136 M
Repo        : installed
Summary     : Manage EC2 Instances using SSM APIs
URL         : http://docs.aws.amazon.com/ssm/latest/APIReference/Welcome.html
License     : Apache License, Version 2.0
Description : This package provides Amazon SSM Agent for managing EC2 Instances
            : using SSM APIs


# file asa.rpm
asa.rpm: RPM v3.0 bin i386/x86_64 amazon-ssm-agent-2.3.1644.0-1

**Comment by relrod** [comment]
I looked into this briefly today - `yum` does return 1 in this case, so we are currently staying true to what yum does. Whether or not we should or should not, is another question.

I don't think solving this will be as simple as a substring search in the result because we'd have to deal with the case of acting on multiple packages and only one of them hitting this condition, etc.

CCing @maxamillion who can weigh in more.

**Comment by maxamillion** [comment]
This is a really weird one and I'm surprised the `yum` cli does in fact return 1. I am on the fence here, I agree we're sticking true to what yum says *but* we're violating the idempotent nature of the module which I think is a bug. Let me try and find some time and dig into this a bit to see if we can catch this corner case properly.

**Comment by maxamillion** [comment]
Alright so on second look of this we're not violating the idempotent nature since you can alternatively specify `allow_downgrade: yes` in the task and it will perform the appropriate action just as YUM itself will (and this parameter appears to be required to the YUM command line to mimic the desired behavior also. Therefore I don't think this a bug, but functioning as expected (since we're inheriting behavior from YUM in this instance).

cc @relrod 

I'm going to go ahead and close this.

**Comment by chris-kiick-sp** [comment]
I disagree.  Since the state was specified as “latest” it should either install the package, if it is newer, or succeed with changed=false.
Using the downgrade flag does restore the idempotent behavior, but does not retain the ability to install or keep the newer package, which is what the “latest” state is for.

This could be documented as an exception, but I think explaining it in the docs would take about as much room as the rest of it.

I have found a work-around using failed_when and changed_when, for this case, so I’m not blocked on this.  However, surely someone else is going to run into this and call the behavior unexpected.

--
Chris J. Kiick
Sr. Cloud Performance Engineer
chris.kiick@sailpoint.com<mailto:chris.kiick@sailpoint.com>
www.sailpoint.com<http://www.sailpoint.com>

**Comment by maxamillion** [comment]
@chris-kiick-sp I think I misunderstood, I am getting `changed=false` when I run this against a machine where the package is not an upgrade using Ansible 2.9.13 ... I'm happy to re-evaluate and/or learn more about the issue you're facing but I setup a local VM with RHEL 7.8 and tested against latest stable (Anisble 2.9.13 at the time of my testing ... but 2.10.0 is out now) and it worked as expected.

**Comment by maxamillion** [comment]
@chris-kiick-sp I'll reopen the issue so we can continue the discussion here.

**Comment by s-hertel** [comment]
This is fixed on 2.10, so I'm closing it as per above. Please open a new issue with reproducer if you're experiencing this on recent versions.
